File: internal/plan/progress.go
package plan

import (
	"encoding/json"
	"os"
	"path/filepath"
	"time"
)

const progressLogFileName = "progress.log"

// Event type constants for progress logging.
const (
	EventPlanStarted   = "plan_started"
	EventPlanCompleted = "plan_completed"
	EventPlanCancelled = "plan_cancelled"
	EventPlanFailed    = "plan_failed"
	EventTaskStarted   = "task_started"
	EventTaskCompleted = "task_completed"
	EventTaskFailed    = "task_failed"
)

// ProgressEvent represents a single progress log entry.
type ProgressEvent struct {
	Timestamp time.Time              `json:"timestamp"`
	Event     string                 `json:"event"`
	Data      map[string]interface{} `json:"data,omitempty"`
}

// ProgressLogger writes progress events to a JSON Lines file.
type ProgressLogger struct {
	path string
}

// NewProgressLogger creates a new progress logger for the given plan directory.
func NewProgressLogger(planDir string) *ProgressLogger {
	return &ProgressLogger{
		path: filepath.Join(planDir, progressLogFileName),
	}
}

// Log appends a progress event to the log file.
func (p *ProgressLogger) Log(event string, data map[string]interface{}) error {
	entry := ProgressEvent{
		Timestamp: time.Now(),
		Event:     event,
		Data:      data,
	}

	jsonBytes, err := json.Marshal(entry)
	if err != nil {
		return err
	}
	jsonBytes = append(jsonBytes, '\n')

	f, err := os.OpenFile(p.path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return err
	}
	defer f.Close()

	_, err = f.Write(jsonBytes)
	return err
}

// PlanStarted logs a plan_started event.
func (p *ProgressLogger) PlanStarted(planID string) error {
	return p.Log(EventPlanStarted, map[string]interface{}{
		"plan_id": planID,
	})
}

// TaskStarted logs a task_started event.
func (p *ProgressLogger) TaskStarted(taskID string, attempt int) error {
	return p.Log(EventTaskStarted, map[string]interface{}{
		"task_id": taskID,
		"attempt": attempt,
	})
}

// TaskCompleted logs a task_completed event.
func (p *ProgressLogger) TaskCompleted(taskID string) error {
	return p.Log(EventTaskCompleted, map[string]interface{}{
		"task_id": taskID,
	})
}

// TaskFailed logs a task_failed event.
func (p *ProgressLogger) TaskFailed(taskID string, attempt int) error {
	return p.Log(EventTaskFailed, map[string]interface{}{
		"task_id": taskID,
		"attempt": attempt,
	})
}

// PlanCompleted logs a plan_completed event with summary statistics.
func (p *ProgressLogger) PlanCompleted(totalTasks, succeededTasks int, duration time.Duration) error {
	return p.Log(EventPlanCompleted, map[string]interface{}{
		"total_tasks":     totalTasks,
		"succeeded_tasks": succeededTasks,
		"duration_ms":     duration.Milliseconds(),
	})
}

// PlanCancelled logs a plan_cancelled event.
func (p *ProgressLogger) PlanCancelled(lastTaskID string) error {
	return p.Log(EventPlanCancelled, map[string]interface{}{
		"last_task_id": lastTaskID,
	})
}

// PlanFailed logs a plan_failed event.
func (p *ProgressLogger) PlanFailed(taskID string, attempts int) error {
	return p.Log(EventPlanFailed, map[string]interface{}{
		"task_id":  taskID,
		"attempts": attempts,
	})
}

File: ts-port/src/core/progress.ts
/**
 * Progress event logging
 *
 * Implements JSONL progress event logging for plan execution.
 * Events are appended to progress.jsonl in the plan directory.
 */

import * as fs from "node:fs/promises";
import * as path from "node:path";

const PROGRESS_FILE_NAME = "progress.jsonl";

/**
 * Event type constants for progress logging.
 * Matches Go version: internal/plan/progress.go
 */
export const EventType = {
  PlanStarted: "plan_started",
  PlanCompleted: "plan_completed",
  PlanCancelled: "plan_cancelled",
  PlanFailed: "plan_failed",
  TaskStarted: "task_started",
  TaskCompleted: "task_completed",
  TaskFailed: "task_failed",
} as const;

export type ProgressEventType = (typeof EventType)[keyof typeof EventType];

/**
 * Progress event data structure
 */
export interface ProgressEvent {
  timestamp: string;
  event: ProgressEventType;
  data?: Record<string, unknown>;
}

/**
 * ProgressLogger writes progress events to a JSONL file.
 * All writes are append-only.
 */
export class ProgressLogger {
  private readonly filePath: string;

  constructor(planDir: string) {
    this.filePath = path.join(planDir, PROGRESS_FILE_NAME);
  }

  /**
   * Appends a progress event to the log file.
   */
  async log(
    event: ProgressEventType,
    data?: Record<string, unknown>
  ): Promise<void> {
    const entry: ProgressEvent = {
      timestamp: new Date().toISOString(),
      event,
      data,
    };

    const jsonLine = JSON.stringify(entry) + "\n";

    // Open file in append mode, create if not exists
    const handle = await fs.open(this.filePath, "a");
    try {
      await handle.write(jsonLine);
    } finally {
      await handle.close();
    }
  }

  /**
   * Logs a plan_started event.
   */
  async planStarted(planId: string): Promise<void> {
    return this.log(EventType.PlanStarted, { plan_id: planId });
  }

  /**
   * Logs a task_started event.
   */
  async taskStarted(taskId: string, attempt: number): Promise<void> {
    return this.log(EventType.TaskStarted, { task_id: taskId, attempt });
  }

  /**
   * Logs a task_completed event.
   */
  async taskCompleted(taskId: string): Promise<void> {
    return this.log(EventType.TaskCompleted, { task_id: taskId });
  }

  /**
   * Logs a task_failed event.
   */
  async taskFailed(taskId: string, attempt: number): Promise<void> {
    return this.log(EventType.TaskFailed, { task_id: taskId, attempt });
  }

  /**
   * Logs a plan_completed event with summary statistics.
   */
  async planCompleted(
    totalTasks: number,
    succeededTasks: number,
    durationMs: number
  ): Promise<void> {
    return this.log(EventType.PlanCompleted, {
      total_tasks: totalTasks,
      succeeded_tasks: succeededTasks,
      duration_ms: durationMs,
    });
  }

  /**
   * Logs a plan_cancelled event.
   */
  async planCancelled(lastTaskId: string): Promise<void> {
    return this.log(EventType.PlanCancelled, { last_task_id: lastTaskId });
  }

  /**
   * Logs a plan_failed event.
   */
  async planFailed(taskId: string, attempts: number): Promise<void> {
    return this.log(EventType.PlanFailed, { task_id: taskId, attempts });
  }

  /**
   * Gets the path to the progress file.
   */
  getFilePath(): string {
    return this.filePath;
  }
}
